lms/djangoapps/course_blocks/transformers/split_test.py

```diff
@@ -0,0 +1,110 @@
+            user_groups (dict[Partition Id: Group])
+
+        Returns:
+            bool
+        """
+        if split_test_groups:
+            for _, group in user_groups.iteritems():
+                if group.id in split_test_groups:
+                    return True
+            return False
+        return True
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Computes any information for each XBlock that's necessary to execute
+        this transformer's transform method.
+
+        Arguments:
+            block_structure (BlockStructureCollectedData)
+        """
+
+        # Check potential previously set values for user_partitions and split_test_partitions
+        xblock = block_structure.get_xblock(block_structure.root_block_key)
+        user_partitions = getattr(xblock, 'user_partitions', [])
+        split_test_partitions = getattr(xblock, 'split_test_partition', []) or []
+
+        # For each block, check if it is a split_test block.
+        # If split_test is found, check its user_partition value and get children.
+        # Set split_test_group on each of the children for fast retrieval in transform phase.
+        # Add same group to children's children, because due to structure restrictions first level
+        # children are verticals.
+        for block_key in block_structure.topological_traversal():
+            xblock = block_structure.get_xblock(block_key)
+            category = getattr(xblock, 'category', None)
+            if category == 'split_test':
+                for user_partition in user_partitions:
+                    if user_partition.id == xblock.user_partition_id:
+                        if user_partition not in split_test_partitions:
+                            split_test_partitions.append(user_partition)
+                        for child in xblock.children:
+                            for group in user_partition.groups:
+                                child_location = xblock.group_id_to_child.get(
+                                    unicode(group.id),
+                                    None
+                                )
+                                if child_location == child:
+                                    block_structure.set_transformer_block_data(
+                                        child,
+                                        cls,
+                                        'split_test_groups',
+                                        [group.id]
+                                    )
+                                    for component in block_structure.get_xblock(child).children:
+                                        block_structure.set_transformer_block_data(
+                                            component,
+                                            cls,
+                                            'split_test_groups',
+                                            [group.id]
+                                        )
```

Comment:
I'm sorry if this was not clarified earlier.  We need to support:
* nested SplitTest modules (you can theoretically have split_tests within split_tests)
* a block hierarchy of arbitrary depth within a split_test module (cannot assume just upto a grandchild)
* if possible, DAGs (a block can have multiple parents)

Given this, I would implement this by storing the intersection of all user_partitions in a parent-chain, unioned with all parent-chains (in case of multiple parents).  However, this is exactly what has already been implemented in the UserPartitionsTransformer.  So I'm wondering whether we can just rely on that transformer.  I am looking into this now.